src/crimson/net/Session.h

```diff
@@ -0,0 +1,106 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel Corp.
+ *
+ * Author: Yingxin Cheng <yingxincheng@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+
+#include "Fwd.h"
+#include "Socket.h"
+
+#include "crimson/thread/Throttle.h"
+
+#include "include/random.h"
+#include "auth/AuthSessionHandler.h"
+#include "msg/Policy.h"
+
+namespace ceph::net {
+
+struct Session
+{
+  entity_addr_t my_addr;
+  peer_type_t peer_type;
+  entity_addr_t peer_addr;
+  std::chrono::milliseconds backoff;
+  uint32_t connect_seq = 0;
+  uint32_t peer_global_seq = 0;
+  uint32_t global_seq;
+  ceph_timespec last_keepalive_ack;
+
+  std::optional<Socket> socket;
+
+  seastar::promise<> send_promise;
+
+  ceph::net::Policy<ceph::thread::Throttle> policy;
+
+  uint64_t features;
+  /// the seq num of the last transmitted message
+  seq_num_t out_seq = 0;
+  /// the seq num of the last received message
+  seq_num_t in_seq = 0;
+
+  std::unique_ptr<AuthSessionHandler> session_security;
+  // messages to be resent after connection gets reset
+  std::queue<MessageRef> out_q;
+  // messages sent, but not yet acked by peer
+  std::queue<MessageRef> sent;
+
+  seastar::gate dispatch_gate;
+
+  Session(const entity_addr_t& _my_addr,
+          const entity_addr_t& _peer_addr)
+    : my_addr(_my_addr), peer_addr(_peer_addr) {}
+
+  Session(const entity_addr_t& _my_addr,
+          const entity_addr_t& _peer_addr,
+          seastar::connected_socket&& fd)
+    : my_addr(_my_addr), peer_addr(_peer_addr),
+      socket(std::forward<seastar::connected_socket>(fd)) {}
```

Comment:
nit, why not just use `std::move()` here?

Reply:
It is to use perfect forwarding to keep `fd` as an rvalue. Do you think it is necessary here?